### ISSUE TITLE
[java] ModifierOrder false-positives with type annotations and type parameters (typeAnnotations = anywhere)

### ISSUE THREAD

**Issue by jjlharrison** [pmd/pmd#6053]
**Affects PMD Version:** 

7.17.0

**Rule:** [ModifierOrder](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#modifierorder)

**Description:**

**Code Sample demonstrating the issue:**

```xml
<rule ref="category/java/codestyle.xml/ModifierOrder">
    <properties>
        <property name="typeAnnotations" value="anywhere" />
    </properties>
</rule>
```

**Case 1:**

```java
package com.marshmallow.example.springdoc;

import io.swagger.v3.oas.models.ExternalDocumentation;
import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.info.Contact;
import io.swagger.v3.oas.models.info.Info;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.lang.NonNull;
import org.springframework.web.servlet.config.annotation.CorsRegistry;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

@TestConfiguration
public class ApiDocsConfiguration {

    @Bean
    public OpenAPI openApi() {
        return new OpenAPI()
            .externalDocs(new ExternalDocumentation()
                .url("https://github.com/marshmallow-insurance/uk-auto-platform-sandbox/blob/main/README.md")
                .description("README.md")
            )
            .info(new Info()
                .title("Example Service API")
                .contact(new Contact().name("Multi Product").url("https://app.opslevel.com/teams/multi_product"))
                .description("An example configuration to test integration between Spring and Springdoc dependencies.")
            );
    }

    @Bean
    public WebMvcConfigurer corsConfigurer() {
        return new WebMvcConfigurer() {
            @Override
            public void addCorsMappings(final @NonNull CorsRegistry registry) { // <= False positive here
                registry.addMapping("/**")
                    .allowedOrigins("https://docs.marshmallow.com")
                    .allowCredentials(true);
            }
        };
    }

}
```

**Case 2:**
https://github.com/pmd/pmd/blob/2dffbed5a4b0a95b321273bbecca88430372f5db/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/internal/StreamImpl.java#L225

**Expected outcome:**

Case 1: ~PMD reports a violation at line 34, but as we allow type annotations anywhere, this should not be reported.~
Correct behavior of the rule: org.springframework.lang.NonNull is not a type annotation and should therefore be placed before keywords.

Case 2: PMD reports a violation at line 225, but as we allow type annotations anywhere, this should not be reported.

**Running PMD through:** Maven

```
[INFO] --- pmd:3.27.0:pmd (pmd) @ example-service ---
[INFO] PMD version: 7.17.0
[INFO] Rendering content with org.apache.maven.skins:maven-fluido-skin:jar:2.0.0-M9 skin
[INFO]
[INFO] <<< pmd:3.27.0:check (pmd) < :pmd @ example-service <<<
[INFO]
[INFO]
[INFO] --- pmd:3.27.0:check (pmd) @ example-service ---
[WARNING] PMD Failure: com/marshmallow/example/springdoc/ApiDocsConfiguration.java:34 Rule:ModifierOrder Priority:1 Missorted modifiers `final @NonNull`. Annotations should be placed before modifiers..
```

**Comment by bmeier-pros** [comment]
I also hit this with the default configuration of the ModifierOrder.

**Comment by adangel** [comment]
@jjlharrison I've updated the issue description a bit:

Case 1: This is actually correct behavior. Spring's [NonNull](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/NonNull.html) annotation is not a type annotation and the rule enforces, that non-type annotations should be before the keywords, regardless of the typeAnnotation property:
> Annotations that are not type annotations are still required to be before keyword modifiers.
> https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#modifierorder


Case 2: Is a real false-positive, as type parameters weren't considered yet and then type annotations weren't correctly detected.

**Comment by sgoryachkin** [comment]
I have the same problem with @ JsonView (Jackson annotation). It is a type annotation

<img width="819" height="626" alt="Image" src="https://github.com/user-attachments/assets/cff085c1-e981-4940-8f88-09a784c641d3" />

[WARNING] PMD Failure: p.k.g.MyController:57 Rule:ModifierOrder Priority:1 Missorted modifiers `public @ JsonView(...)`. Annotations should be placed before modifiers..

I use default configuration of the rule
PMD 7.17, 7.18

**Comment by adangel** [comment]
@sgoryachkin Can you show the part, where the rule is triggering? Are there any generics in this line of code in addition to the annotation? Or in other words: Is your case equivalent to case 2 or is it a new case?

**Comment by sgoryachkin** [comment]
> [@sgoryachkin](https://github.com/sgoryachkin) Can you show the part, where the rule is triggering? Are there any generics in this line of code in addition to the annotation? Or in other words: Is your case equivalent to case 2 or is it a new case?

I tried deleting all other annotations. I left only com.fasterxml.jackson.annotation.JsonView

<img width="1267" height="143" alt="Image" src="https://github.com/user-attachments/assets/74239516-869a-4f6c-bf2a-f41767cd2b05" />

The rule (default configuration) is triggered on line 43. PMD 7.18, pmd-maven-plugin 3.28, jdk 21

<img width="1903" height="36" alt="Image" src="https://github.com/user-attachments/assets/0feb938f-6e91-4aaf-b9fd-5e5cb4b7d256" />

**Comment by adangel** [comment]
>> Are there any generics

Yes: `public @JsonView(ResponseView.class) ResponseEntity<AbstractTask> create(...) {}`

@sgoryachkin I assume, that your case is equivalent to case 2 and should also be fixed with PMD 7.19.0 (due tomorrow). If it not fixed, please open a new issue, thanks!

**Comment by adangel** [comment]
@sgoryachkin I actually tested your case. The annotation JsonView is _not_ a type annotation, means it does _not_ annotate the return type of the create() method. Instead it annotations the method itself (`ElementType.METHOD`). `ElementType.TYPE` allows the annotation to be placed on a class declaration. And `ElementType.TYPE_USE` (which is missing for JsonView) would allow to place it on a return type.

So I think, the rule is correct to demand to put the JsonView annotation before the "public" modifier to make it clear, it annotates the method.
